File: 프로그래머스/2/150369. 택배 배달과 수거하기/택배 배달과 수거하기.py
from math import ceil

def solution(cap, n, deliveries, pickups):
    answer = 0
    deliver_point, pickup_point = [], []
    remain_d, remain_p = 0, 0
    for i in range(n-1, -1, -1):
        if deliveries[i] != 0: 
            count = ceil((deliveries[i] - ((cap -remain_d)%cap)) / cap)
            remain_d = (deliveries[i] - ((cap -remain_d)%cap)) % cap
            deliver_point += [i] * count

        if pickups[i] != 0: 
            count = ceil((pickups[i] - ((cap -remain_p)%cap)) / cap)
            remain_p = (pickups[i] - ((cap -remain_p)%cap)) % cap
            pickup_point += [i] * count


    if len(deliver_point) == len(pickup_point):
        for i in range(len(pickup_point)):
            answer += (max(deliver_point[i], pickup_point[i]) +1)*2
    elif len(deliver_point) < len(pickup_point):
        for i in range(len(pickup_point)):
            try:
                answer += (max(deliver_point[i], pickup_point[i])+1)*2
            except:
                answer += (pickup_point[i] +1)*2
    else:
        for i in range(len(deliver_point)):
            try:
                answer += (max(deliver_point[i], pickup_point[i])+1)*2
            except:
                answer += (deliver_point[i] +1)*2
    return answer

File: 프로그래머스/2/150369. 택배 배달과 수거하기/택배 배달과 수거하기.java
class Solution {
    static int [] deliver, pickup;
    public long solution(int cap, int n, int[] deliveries, int[] pickups) {
        deliver = deliveries.clone(); pickup = pickups.clone();
        long answer = 0;
        int maxD = -1, maxP = -1;
        for (int i = n-1; i > -1; i--){
            if (maxD == -1 && deliveries[i] > 0) maxD = i;
            if (maxP == -1 && pickups[i] > 0) maxP = i;
        }
        int dst;
        
        while (maxD > -1 || maxP > -1){
            dst = Math.max(maxD, maxP);
            answer += (dst+1) * 2;
            if (maxD > -1)
                maxD = update(cap, maxD, 1);
            if (maxP > -1)
                maxP = update(cap, maxP, 0);
        }
        
        
        return answer;
    }
    
    static int update(int amt, int maxD, int type){
        if (type == 1){
            while (maxD>-1) {
                if (deliver[maxD] > amt) {
                    deliver[maxD] -= amt;
                    amt = 0;
                    break;
                }
                else if (deliver[maxD] <= 0) {
                    maxD--;
                }
                else {
                    amt -= deliver[maxD];
                    deliver[maxD] = 0;
                    maxD--;
                }
            }
        }
        else {
            while (maxD >-1) {
                if (pickup[maxD] > amt) {
                    pickup[maxD] -= amt;
                    amt = 0;
                    break;
                }
                else if (pickup[maxD] <= 0) {
                    maxD--;
                }
                else {
                    amt -= pickup[maxD];
                    pickup[maxD] = 0;
                    maxD--;
                }
            }
        }
        return maxD;
    }
}